File: antob/compilation.py
# compilation.py
import os
import re
import subprocess
from utility import parse_errors
from conversion import fix_cs

# function to execute c# compiler

from subprocess import check_output, CalledProcessError
def compile_cs(path):    
    try:
        output = check_output(['csc', path])
        if not output.strip():
            return []
        else:
            return [line for line in output.decode().split('\n') if 'error' in line]
    except CalledProcessError as e:
        return ['Compilation failed with exit code {}'.format(e.returncode)]


def compile_project(dest_dir):
    # Define the command
    command = "dotnet build"

    # Call the command in the specified directory
    process = subprocess.Popen(command, shell=True, cwd=dest_dir, stdout=subprocess.PIPE, stderr=subprocess.PIPE)

    # Wait for the command to complete
    stdout, stderr = process.communicate()

    # Check if the command was successful
    if process.returncode == 0:
        print("Build succeeded.")
        return None
    else:
        print("Build failed. Errors:")
        errors = parse_errors(stdout.decode())
        for file, error_list in errors.items():
            print(f"{file}:")
            for error in error_list:
                print(f"  {error}")
        return errors

def fix_errors(errors, dest_dir,tokenizer,model):
    # ask model to fix errors
    for file, error_list in errors.items():
       if file.endswith(".razor.cs"):
          print(f"Fixing errors in {file}:")
          with open(file, 'r', encoding='utf-8') as f:
             csharp = f.read()
             error_string = '\n  '.join(error_list)
             fixed = fix_cs(dest_dir,csharp,error_string,tokenizer,model)
             with open(file, 'w', encoding='utf-8') as f:
                f.write(fixed)

File: csharp/ngtob/Compilation.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;

public static class Compilation
{
    public static List<string> CompileCs(string path)
    {
        List<string> errors = new List<string>();
        try
        {
            ProcessStartInfo psi = new ProcessStartInfo
            {
                FileName = "csc",
                Arguments = path,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            using (Process process = Process.Start(psi))
            {
                process.WaitForExit();
                string output = process.StandardOutput.ReadToEnd();
                string errorOutput = process.StandardError.ReadToEnd();
                string combinedOutput = output + "\n" + errorOutput;
                if (!string.IsNullOrWhiteSpace(output))
                {
                    errors = combinedOutput.Split(new[] { '\n' }, StringSplitOptions.RemoveEmptyEntries)
                                           .Where(line => line.IndexOf("error", StringComparison.OrdinalIgnoreCase) >= 0)
                                           .ToList();
                }
            }
        }
        catch (Exception ex)
        {
            errors.Add("Compilation failed with exception: " + ex.Message);
        }
        return errors;
    }

    public static Dictionary<string, List<string>> CompileProject(string destDir)
    {
        string command = "dotnet build";

        ProcessStartInfo psi = new ProcessStartInfo
        {
            FileName = "cmd.exe",
            Arguments = $"/c {command}",
            WorkingDirectory = destDir,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        };

        Dictionary<string, List<string>> errors = new Dictionary<string, List<string>>();
        using (Process process = Process.Start(psi))
        {
            process.WaitForExit();
            string stdout = process.StandardOutput.ReadToEnd();
            string stderr = process.StandardError.ReadToEnd();
            if (process.ExitCode == 0)
            {
                Console.WriteLine("Build succeeded.");
                return null;
            }
            else
            {
                Console.WriteLine("Build failed. Errors:");
                errors = Utility.ParseErrors(stdout);
                foreach (var kvp in errors)
                {
                    Console.WriteLine($"{kvp.Key}:");
                    foreach (var error in kvp.Value)
                    {
                        Console.WriteLine($"  {error}");
                    }
                }
            }
        }
        return errors;
    }

    public static void FixErrors(Dictionary<string, List<string>> errors, string destDir)
    {
        // For each file with errors, call FixCs to get a fixed version.
        foreach (var kvp in errors)
        {
            string file = kvp.Key;
            if (file.EndsWith(".razor.cs", StringComparison.OrdinalIgnoreCase))
            {
                Console.WriteLine("Fixing errors in " + file);
                string csharp = File.ReadAllText(file);
                string errorString = string.Join("\n  ", kvp.Value);
                string fixedCode = Conversion.FixCs(destDir, csharp, errorString);
                File.WriteAllText(file, fixedCode);
            }
        }
    }
}
